### PR TITLE
Add key used on lookup to head transmitted through proxy and emit head too

### PR DESCRIPTION
**Pull request by jwolski**
This is a breaking change required for serial processing

@Raynos

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -77,9 +78,16 @@ var InvalidLocalMemberError = TypedError({
 });
 
 var OptionsRequiredError = TypedError({
-    type: 'ringpop.options.required',
+    type: 'ringpop.options-required',
```
Comment (Raynos) [inline index.js]: these type strings cannot be changed. sorry.

Comment (jwolski) [inline index.js]: DDDD:

File: lib/request-proxy.js [hunk 2]
```diff
@@ -44,15 +44,21 @@ function RequestProxy(ringpop) {
 
 var proto = RequestProxy.prototype;
 
-proto.proxyReq = function proxyReq(dest, req, res, opts) {
+proto.proxyReq = function proxyReq(opts) {
+    var key = opts.key;
+    var dest = opts.dest;
+    var req = opts.req;
+    var res = opts.res;
+    var opts2 = opts.opts;
```
Comment (Raynos) [inline lib/request-proxy.js]: this is weird.

Comment (Raynos) [inline lib/request-proxy.js]: maybe put all the other opts keys flatly onto opts ?

Comment (jwolski) [inline lib/request-proxy.js]: yes.

File: index.js [hunk 3]
```diff
@@ -807,8 +826,24 @@ RingPop.prototype.handleOrProxy =
         if (this.whoami() === dest) {
             return true;
         } else {
-            this.proxyReq(dest, req, res, opts);
+            this.proxyReq({
```
Comment (Raynos) [inline index.js]: could do `extend(opts, { key, dest, req, res })`

File: index.js [hunk 4]
```diff
@@ -19,6 +19,7 @@
 // THE SOFTWARE.
 'use strict';
 
+var _ = require('underscore');
```
Comment (Raynos) [inline index.js]: nitpick: `var extend = require('xtend')`

File: index.js [hunk 5]
```diff
@@ -807,8 +821,23 @@ RingPop.prototype.handleOrProxy =
         if (this.whoami() === dest) {
             return true;
         } else {
-            this.proxyReq(dest, req, res, opts);
+            this.proxyReq(_.extend(opts, {
```
Comment (Raynos) [inline index.js]: would advice against a mutable extension. This might cause suprises for the caller.

File: lib/request-proxy.js [hunk 6]
```diff
@@ -70,7 +75,10 @@ proto.proxyReq = function proxyReq(dest, req, res, opts) {
             headers: headers,
             method: method,
             httpVersion: httpVersion,
-            checksum: ringpop.membership.checksum
+            checksum: ringpop.membership.checksum,
+            ringpopKey: key,
+            ringpopSource: ringpop.whoami(),
+            ringpopDest: dest
```
Comment (Raynos) [inline lib/request-proxy.js]: dest is not needed as it's available as `hostInfo` on the remote side.

Let's not send it unless you want to do packet inspection or something.

File: lib/request-proxy.js [hunk 7]
```diff
@@ -70,7 +75,10 @@ proto.proxyReq = function proxyReq(dest, req, res, opts) {
             headers: headers,
             method: method,
             httpVersion: httpVersion,
-            checksum: ringpop.membership.checksum
+            checksum: ringpop.membership.checksum,
+            ringpopKey: key,
+            ringpopSource: ringpop.whoami(),
```
Comment (Raynos) [inline lib/request-proxy.js]: source is also duplicate information that is available as `hostInfo`

**Comment by Raynos** [comment]
Let's fix:

 - do not mutate arguments.

Otherwise, lgtm.

File: index.js (hunk 5 above)
Comment (jwolski) [inline index.js]: yes, definitely. for some reason i thought it wouldn't mutate `opts` and would generate a copy.

Comment (Raynos) [inline index.js]: `xtend` generates a copy by default ( sane defaults ). `_.extend` mutates by default.

File: index.js (hunk 4 above)
Comment (jwolski) [inline index.js]: no

File: lib/request-proxy.js (hunk 6 above)
Comment (jwolski) [inline lib/request-proxy.js]: we will eventually do tchannel snooping with Matt's tchannel's packet capturer. ill leave it out for now until we need it.

**Pull request merged**